<HTML>
<HEAD>
<TITLE>GAL::Schema::Result::Feature</TITLE>
</HEAD>
<BODY BGCOLOR="#ffffff" TEXT="#000000">
<UL>
<LI><A HREF="#NAME">NAME

</A><LI><A HREF="#VERSION">VERSION

</A><LI><A HREF="#SYNOPSIS">SYNOPSIS

</A><LI><A HREF="#DESCRIPTION">DESCRIPTION

</A><LI><A HREF="#METHODS">METHODS

</A><UL>
<LI><A HREF="#seqid">seqid

</A><LI><A HREF="#source">source

</A><LI><A HREF="#type">type

</A><LI><A HREF="#start">start

</A><LI><A HREF="#end">end

</A><LI><A HREF="#score">score

</A><LI><A HREF="#strand">strand

</A><LI><A HREF="#phase">phase

</A><LI><A HREF="#attributes_hash">attributes_hash

</A><LI><A HREF="#attribute_value">attribute_value

</A><LI><A HREF="#feature_seq">feature_seq

</A><LI><A HREF="#seq">seq

</A><LI><A HREF="#gc_content">gc_content

</A><LI><A HREF="#genomic_seq">genomic_seq

</A><LI><A HREF="#length">length

</A><LI><A HREF="#my_start">my_start

</A><LI><A HREF="#my_end">my_end

</A><LI><A HREF="#genomic_length">genomic_length

</A><LI><A HREF="#annotation">annotation

</A><LI><A HREF="#throw">throw

</A><LI><A HREF="#warn">warn

</A><LI><A HREF="#to_gff3_recursive">to_gff3_recursive

</A><LI><A HREF="#to_gff3">to_gff3

</A><LI><A HREF="#get_recursive_children">get_recursive_children

</A><LI><A HREF="#locus">locus

</A><LI><A HREF="#get_values">get_values

</A><LI><A HREF="#get_transcript_types">get_transcript_types

</A></UL>
<LI><A HREF="#DIAGNOSTICS">DIAGNOSTICS


</A><LI><A HREF="#CONFIGURATION%20AND%20ENVIRONMENT">CONFIGURATION AND ENVIRONMENT

</A><LI><A HREF="#DEPENDENCIES">DEPENDENCIES

</A><LI><A HREF="#INCOMPATIBILITIES">INCOMPATIBILITIES

</A><LI><A HREF="#BUGS%20AND%20LIMITATIONS">BUGS AND LIMITATIONS

</A><LI><A HREF="#AUTHOR">AUTHOR

</A><LI><A HREF="#LICENCE%20AND%20COPYRIGHT">LICENCE AND COPYRIGHT

</A><LI><A HREF="#DISCLAIMER%20OF%20WARRANTY">DISCLAIMER OF WARRANTY

</A></UL>
<HR>
<H1><A NAME="NAME">NAME

</A></H1>

<P>GAL::Schema::Result::Feature - Base class for all sequence features

</P><H1><A NAME="VERSION">VERSION

</A></H1>

<P>This document describes GAL::Schema::Result::Feature version 0.01

</P><H1><A NAME="SYNOPSIS">SYNOPSIS

</A></H1>

<PRE>    use GAL::Annotation;
    my $feat_store = GAL::Annotation-&gt;new(storage =&gt; $feat_store_args,
					  parser  =&gt; $parser_args,
					  fasta   =&gt; $fasta_args,
					 );

    $feat_store-&gt;load_files(files =&gt; $feature_file,
			    mode  =&gt; 'overwrite',
			    );

    my $features = $feat_store-&gt;schema-&gt;resultset('Feature');

    my $mrnas = $features-&gt;search({type =&gt; 'mRNA'});
    while (my $mrna = $mrnas-&gt;next) {
      my $cdss = $mrna-&gt;CDSs;
      while (my $cds = $cdss-&gt;next) {
	my $cds   = $cds-&gt;feature_id;
	my $start = $cds-&gt;start;
	my $end   = $cds-&gt;end;
      }
    }</PRE>
<H1><A NAME="DESCRIPTION">DESCRIPTION

</A></H1>

<P>&lt;GAL::Schema::Result::Feature&gt; is the base class for all sequence
feature classes in the GAL's &lt;DBIx::Class&gt; based feature retrival
system.  You don't instantiate objects from it or any of it's
subclasses yourself, &lt;DBIx::Class&gt; does that for you, hence there is
no constructor and there are no attributes available, but you will
want to have a look at the methods as they provide the base
functionality for all sequence feature subclasses.  to all subclasses
as well.  The first 8 methods provide access to the features primary
data as defined by the first 8 columns of GFF3
(http://www.sequenceontology.org/).

</P><H1><A NAME="METHODS">METHODS

</A></H1>
<H2><A NAME="seqid">seqid

</A></H2>

<PRE> Title   : seqid
 Usage   : $seqid = $self-&gt;seqid
 Function: Get the features sequence ID (i.e. what chromosome/contig it's on).
 Returns : The value of the sequence ID as text.
 Args    : None</PRE>
<H2><A NAME="source">source

</A></H2>

<PRE> Title   : source
 Usage   : $source = $self-&gt;source
 Function: Get the features source.
 Returns : The value of the source as text.
 Args    : None</PRE>
<H2><A NAME="type">type

</A></H2>

<PRE> Title   : type
 Usage   : $type = $self-&gt;type
 Function: Get the features type - a value constrained to be a child of the
	   SO's sequence feature term.
 Returns : The value of the type as text.
 Args    : None</PRE>
<H2><A NAME="start">start

</A></H2>

<PRE> Title   : start
 Usage   : $start = $self-&gt;start
 Function: Get the features start on the sequence described by seqid.
 Returns : The value of the start an integer.
 Args    : None</PRE>
<H2><A NAME="end">end

</A></H2>

<PRE> Title   : end
 Usage   : $end = $self-&gt;end
 Function: Get the features end on the sequence described by seqid.
 Returns : The value of the end as an integer.
 Args    : None</PRE>
<H2><A NAME="score">score

</A></H2>

<PRE> Title   : score
 Usage   : $score = $self-&gt;score
 Function: Get the features score.
 Returns : The value of the score as text.
 Args    : None</PRE>
<H2><A NAME="strand">strand

</A></H2>

<PRE> Title   : strand
 Usage   : $strand = $self-&gt;strand
 Function: Get the features strand.
 Returns : The value of the strand as text.
 Args    : None</PRE>
<H2><A NAME="phase">phase

</A></H2>

<PRE> Title   : phase
 Usage   : $phase = $self-&gt;phase
 Function: Get the features phase.
 Returns : The value of the phase as text.
 Args    : None</PRE>
<H2><A NAME="attributes_hash">attributes_hash

</A></H2>

<PRE> Title   : attributes_hash
 Usage   : $self-&gt;attributes_hash
 Function: Return the attributes as a hash (or reference) with all values as
	   array references.  For consistency, even those values, such as ID,
	   that can have only one value are still returned as array
	   references.
 Returns : A hash or hash reference of attribute key/value pairs.
 Args    : None.</PRE>
<H2><A NAME="attribute_value">attribute_value

</A></H2>

<PRE> Title   : attribute_value
 Usage   : $self-&gt;attribute_value($tag)
 Function: Return the value(s) of a particular attribute as an array or
	   reference.  Note that for consistency, values are always returned
	   as arrays (or reference) even in cases where only a single value
	   could exist such as ID.
 Returns : An array or reference of values.
 Args    : None</PRE>
<H2><A NAME="feature_seq">feature_seq

</A></H2>

<PRE> Title   : feature_seq
 Usage   : $self-&gt;feature_seq
 Function: Returns the features sequence as text 5' to 3' in the context of
	   the feature, so features on the reverse strand are reverse
	   complimented.
 Returns : A text string of the features sequence.
 Args    : None</PRE>
<H2><A NAME="seq">seq

</A></H2>

<P>An alias for feature_seq

</P><H2><A NAME="gc_content">gc_content

</A></H2>

<PRE> Title   : gc_content
 Usage   : $self-&gt;gc_content
 Function: Returns the ratio of the count of G or C nts over the sequence
	   length.
 Returns : A fractional value from 0 to 1
 Args    : None</PRE>
<H2><A NAME="genomic_seq">genomic_seq

</A></H2>

<PRE> Title   : genomic_seq
 Usage   : $self-&gt;genomic_seq
 Function: Returns the features sequence as text 5' to 3' in the context of
	   the genome, so features on the reverse strand are not reverse
	   complimented.
 Returns : A text string of the features sequence on the genome.
 Args    : None</PRE>
<H2><A NAME="length">length

</A></H2>

<PRE> Title   : length
 Usage   : $self-&gt;length
 Function: Returns the features length on the genome.  If a feature is
	   not contiguous on the genome (i.e. a transcript) then this method
	   will be (or should be) overridden by that subclass to provide the
	   'spliced' length of that feature.
 Returns : An integer.
 Args    : None</PRE>
<H2><A NAME="my_start">my_start

</A></H2>

<PRE> Title   : my_start
 Usage   : $self-&gt;my_start
 Function: Returns the begin 5' coordinate on the feature's strand
	    (i.e. this is the same as calling start for a + strand
	    feature and end for a - strand feature.
 Returns : An integer.
 Args    : None</PRE>
<H2><A NAME="my_end">my_end

</A></H2>

<PRE> Title   : my_end
 Usage   : $self-&gt;my_end
 Function: Returns the begin 5' coordinate on the feature's strand
	    (i.e. this is the same as calling end for a + strand
	    feature and start for a - strand feature.
 Returns : An integer.
 Args    : None</PRE>
<H2><A NAME="genomic_length">genomic_length

</A></H2>

<PRE> Title   : genomic_length
 Usage   : $self-&gt;genomic_length
 Function: Returns the features genomic_length on the genome.  If a feature is
	   not contiguous on the genome (i.e. a transcript) then this method
	   will still provide the genomic length of (end - start + 1)
 Returns : An integer.
 Args    : None</PRE>
<H2><A NAME="annotation">annotation

</A></H2>

<PRE> Title   : annotation
 Usage   : $self-&gt;annotation
 Function: Each feature has a weakened reference of the GAL::Annotation
	   object that created it and this method provides access to that
	   object.
 Returns : A GAL::Annotation object.
 Args    : None</PRE>
<H2><A NAME="throw">throw

</A></H2>

<PRE> Title   : throw
 Usage   : $self-&gt;throw
 Function: This is a convinience function that forwards throwings on to the
	   L&lt;GAL::Annotation&gt; object so that L&lt;GAL::Base&gt; can handle them.
	   You could get the same behaviour by doing
	   C&lt;$self-&gt;annotation-&gt;throw(message =&gt; 'throwing'&gt;.
 Returns : Nothing
 Args    : The following list: ($code, $message)</PRE>
<H2><A NAME="warn">warn

</A></H2>

<PRE> Title   : warn
 Usage   : $self-&gt;warn
 Function: This is a convinience function that forwards warnings on to the
	   L&lt;GAL::Annotation&gt; object so that L&lt;GAL::Base&gt; can handle them.
	   You could get the same behaviour by doing
	   C&lt;$self-&gt;annotation-&gt;warn(message =&gt; 'warning'&gt;.
 Returns : Nothing
 Args    : The following list: ($code, $message)</PRE>
<H2><A NAME="to_gff3_recursive">to_gff3_recursive

</A></H2>

<PRE> Title   : to_gff3_recursive
 Usage   : $self-&gt;to_gff3_recursive
 Function: Return a list of the string representation of this feature
	   and all of its part_of children in GFF3 format.
 Returns : An array or ref
 Args    : N/A</PRE>
<H2><A NAME="to_gff3">to_gff3

</A></H2>

<PRE> Title   : to_gff3
 Usage   : $self-&gt;to_gff3
 Function: Return a string representation of this feature in GFF3 format.
 Returns : A string of GFF3 text
 Args    : N/A</PRE>
<H2><A NAME="get_recursive_children">get_recursive_children

</A></H2>

<PRE> Title   : get_recursive_children
 Usage   : $self-&gt;get_recursive_children(\@children);
 Function: Collect all part_of children of an object recursively adding them to
           the provided array reference.
 Returns : Fills the supplied array reference.
 Args    : An array reference.</PRE>
<H2><A NAME="locus">locus

</A></H2>

<PRE> Title   : locus
 Usage   : $locus = $feature-&gt;locus;
 Function: Get the locus of a feature formatted as seqid:start-end
 Returns : A string formatted as seqid:start-end.
 Args    : None</PRE>
<H2><A NAME="get_values">get_values

</A></H2>

<PRE> Title   : get_values
 Usage   : @values = $self-&gt;get_values(qw(seqid start end)
 Function: Get a list of values as an array(ref).
 Returns : An array or ref
 Args : A list of fields to retrieve (seqid source type start end
        score strand phase).</PRE>
<H2><A NAME="get_transcript_types">get_transcript_types

</A></H2>

<PRE> Title   : get_transcript_types
 Usage   : $transcript_types = $self-&gt;get_transcript_types
 Function: Get a list of all features that are types of transcripts.
 Returns : An array or ref
 Args    : None</PRE>
<H1><A NAME="DIAGNOSTICS">DIAGNOSTICS


</A></H1>

<P><A HREF="../../../GAL/Schema/Result/Feature.html">GAL::Schema::Result::Feature</A> does not throw any warnings or errors
at this time.

</P><H1><A NAME="CONFIGURATION%20AND%20ENVIRONMENT">CONFIGURATION AND ENVIRONMENT

</A></H1>

<P>&lt;GAL::Schema::Result::Feature&gt; requires no configuration files or
environment variables.

</P><H1><A NAME="DEPENDENCIES">DEPENDENCIES

</A></H1>

<P>&lt;DBIx::Class&gt;

</P><H1><A NAME="INCOMPATIBILITIES">INCOMPATIBILITIES

</A></H1>

<P>None reported.

</P><H1><A NAME="BUGS%20AND%20LIMITATIONS">BUGS AND LIMITATIONS

</A></H1>

<P>No bugs have been reported.

</P>
<P>Please report any bugs or feature requests to:
barry.moore@genetics.utah.edu

</P><H1><A NAME="AUTHOR">AUTHOR

</A></H1>

<P>Barry Moore &lt;barry.moore@genetics.utah.edu&gt;

</P><H1><A NAME="LICENCE%20AND%20COPYRIGHT">LICENCE AND COPYRIGHT

</A></H1>

<P>Copyright (c) 2012, Barry Moore &lt;barry.moore@genetics.utah.edu&gt;.  All
rights reserved.

</P>
<PRE>    This module is free software; you can redistribute it and/or
    modify it under the same terms as Perl itself.</PRE>
<H1><A NAME="DISCLAIMER%20OF%20WARRANTY">DISCLAIMER OF WARRANTY

</A></H1>

<P>BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE &quot;AS IS&quot; WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

</P>
<P>IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.

</P>
</BODY>
</HTML>
