<HTML>
<HEAD>
<TITLE>The Genome Annotation Library - GAL</TITLE>
</HEAD>
<BODY BGCOLOR="#ffffff" TEXT="#000000">

<h2>The Genome Annotation Library - GAL</h2>

<ul>
  <li>
    <a href="#summary">Summary</a>
  </li>
  <li>
    <a href="#synopsis">Synopsis</a>
  </li>
  <li>
    <a href="#description">Description</a>
  </li>
</ul>

<h3 id="summary">Summary</h3>

<p>
  The Genome Annotation Library (GAL) is a object-oriented Perl
  programming library (and a growing collection of associated
  scripts) that is designed to provide a simple and intuitive
  interface for accomplishing Sequence Ontology empowered analyses of
  genomic annotation data.
</p>

<h3 id="synopsis">Synopsis</h3>

<p>
  A very simple GAL script that infers introns for mRNAs is seen
  here:
</p>

<pre>
  use GAL::Annotation;
  my $features = GAL::Annotation->new($gff_file, $fasta_file)->features;

  # Search for mRNAs and get an iterator;
  my $mrnas = $features->search({type => 'mRNA'});
  print STDERR 'Processing ' . $mrnas->count . ' mRNA features';
  # Iterate mRNAs
  while (my $mrna = $mrnas->next) {
    # Infer (if not existing) and return iterator
    my $introns = $mrna->introns;
    while (my $intron = $introns->next) {
      # Use the intron object for access to data
      my $id    = $intron->feature_id;
      my ($start, $end) = $intron->get_values(qw(start end));
      print $intron->fasta if $intron->gc_content > 50;
      print STDERR join ', ', ($feature_id,
                               $intron->locus,
                               $intron->gc_content
                              );
    }
  }
</pre>

<h3 id="description">Description</h3>

<p>
  GAL provides a collection of object classes for sequence features
  who's inheritance structure is based on the is_a releationships
  between sequence features as described in the Sequence Ontology
  (e.g. an mRNA is_a type of transcript and can thus inherit it's
  methods).  In addition feature classes provide methods which will
  retrieve iterators for child features as described by the SO part_of
  relationships and captured by the GFF3 Parent attribute. For
  example, transcript objects can return iterators for the exons which
  are 'part_of' them.  mRNA objects inherit this behavior to return
  exon iterators from the transcript class and in addition can return
  an iterator for the CDS features that are 'part_of' them.
</p>

<h2 id="Storage">Storage</h2>

<p>
  GAL uses an SQLite relational database backend to store parsed
  sequence features.  This allows flexible searching of sequence
  features and their attributes.  Results of these searches are
  returned as iterators, that, in addition to iterating over the
  returned features, provide useful methods for summarizing details of
  their lists.  This flexibility in searching and random access to the
  data comes at the cost of additional time needed to load, index and
  query the database.  Sometimes, all that you need is the ability to
  parse and iterate over the entire file one feature at a time.  In
  these cases the parser classes can be used directly, as described
  below, and features will be rapidly parsed and returned (in a simple
  hash sturcture rather than feature objects).
</p>

<h2 id="class_structure">Class Structure</h2>

<p>
  GAL uses DBIx::Class to provide sequence feature objects from the
  relational database backend.
</p>

<h2 id="parsers">Parsers</h2>

<p>
  While GAL is specifically designed for parsing GFF3 as the input
  format for seuqence features, an abstract Parser class allows
  straight-forward subclassing such that any input file, where the
  contents can be converted into genome locatable sequence features,
  can can be parsed if an appropriate parser subclass is written.  In
  addition to the default ability to parse GFF3, several other parser
  classes are provided with GAL, and new classes are easily added.
</p>

<h2 id=""></h2>

<p>
</p>

<h3 id='using_gal'>Using GAL</h3>

<p>
  As described above you can use the full functionality of GAL by
  allowing it to load features into a relational database or you can use
  light-weight parser classes directly to rapid sequential access to the
  data in hash structures.
</p>

<p>
  In the first case a GAL::Annotation object is used to access feature
  ojects:
</p>

<pre>
  use GAL::Annotation;
  my $features = GAL::Annotation->new($gff_file, $fasta_file)->features;
  while (my $feature = $features->next) {
      print $feature->feature_id . "\n";
      print $feature->seq . "\n";
  }
</pre>

<p>
  In the second case a GAL::Parser object (or subclass) is used to
  access features as hash data structures:
</p>

<pre>
  use GAL::Parser;
  my $parser = GAL::Parser->new($gff_file);
  while (my $feature = $parser->next_feature_hash) {
      # Not $feature is a hash reference not an object!
      print $feature->{feature_id} . "\n";
      # But this doesn't work $feature->{seq} - no OO sugar.
  }
</pre>

<h3 id='gal_classes'>GAL Classes</h3>

<p>
  The following index can be used to explore the documentation in
  the GAL library.
</p>

<!--
/GAL/Reader/DelimitedLine.html
/GAL/Reader/RecordParser.html
/GAL/Reader/Template.html
/GAL/Reader.html
/GAL/Run.html
-->

<ul>
  <li>
    <a href="GAL/Annotation.html">GAL::Annotation</a> - The primary
    class used to initiate a GAL-based annotation analysis.  Start
    here is you're exploring the library for the first time and want
    access to the full OO feature classes and DB backend.  Provided a
    GFF3 file objects of this class will load the features into a
    database, index the database and can return an iterator object
    based to the features in the data base.
  </li>

  <li>
    <a href="GAL/Base.html">GAL::Base</a> - A base class providing
    shared utility methods for all objects library-wide.  This class
    is not instantiated directly by the user.
  </li>

  <li>
    <a href="GAL/Feature.html">GAL::Feature</a> - Experimental and not
    currently in use.
  </li>

  <!--
  <ul>
    <li>
      <a href="GAL/Feature/sequence_alteration.html">GAL::Feature::sequence_alteration</a>
      - Experimental and not currently in use.
    </li>

    <li>
      <a href="GAL/Feature/sequence_feature.html">GAL::Feature::sequence_feature</a>
      - Experimental and not currently in use.
    </li>
  </ul>
  -->

  <li>
    <a href="GAL/Parser.html">GAL::Parser</a> - Base class for all
    GAL::Parser subclasses.  By default it parsers GFF3/GVF format.
    GAL::Parser is used internally by the GAL::Annotation object to
    parse feature files and can also be used directly to access
    features as hash references ($parser->next_feature_hash).
  </li>

  <ul>
    <li>
      <a href="GAL/Parser/gff3.html">GAL::Parser::gff3</a> - A parser
      class used to parse GFF3 file format.  This is the default
      parser for the library.
    </li>

    <li>
      <a href="GAL/Parser/gff3_fast.html">GAL::Parser::gff3_fast</a>
      - A fast GFF3 parser class that makes the assumption that the
      incoming GFF3 file is well formated and does no validation.  It
      may break in unfreindly ways if the data is not well formated.
    </li>

    <li>
      <a href="GAL/Parser/ncbi_blast_tab.html">GAL::Parser::ncbi_blast_tab.html</a>
      - A parser for NCBI Blast+ output that is formated with the
    </li>

    <li>
      <a href="GAL/Parser/VCFv4_0.html">GAL::Parser::VCFv4_0.html</a>
      - A parser for VCF v4.0 files
    </li>

    <li>
      <a href="GAL/Parser/basic_snp.html">GAL::Parser::basic_snp.html</a>
      - A parser for SNVs in basic tab-delimited format.
    </li>

    <li>
      <a href="GAL/Parser/hgmd_indel.html">GAL::Parser::hgmd_indel.html</a>
      - A parser for indel flat files from HGMD.
    </li>

    <li>
      <a href="GAL/Parser/hgmd_snv.html">GAL::Parser::hgmd_snv.html</a>
      - A parser for SNV flat files from HGMD.
    </li>

    <li>
      <a href="GAL/Parser/maq_cns2snp.html">GAL::Parser::maq_cns2snp.html</a>
      - A parser for MAQ cns2snp output files.
    </li>

    <li>
      <a href="GAL/Parser/samtools_pileup.html">GAL::Parser::samtools_pileup.html</a>
      - A parser for samtools pileup output.
    </li>

    <li>
      <a href="GAL/Parser/ucsc_gene_table.html">GAL::Parser::ucsc_gene_table.html</a>
      - A parser for UCSC refGene (and other) tab-delimited files.
    </li>

    <li>
      <a href="GAL/Parser/ucsc_kg_table.html">::GAL::Parser::ucsc_kg_table.html</a>
      - A parser for UCSC knownGene tab-delimited files.
    </li>

    <li>
      <a href="GAL/Parser/template.html">GAL::Parser::template.html</a>
      - A template to use as a starting point for writing new
      GAL::Parser subclasses for sequence feature formats.
    </li>

    <li>
      <a href="GAL/Parser/template_sequence_alteration.html">GAL::Parser::template_sequence_alteration.html</a>
      - A template to use as a starting point for writing new
      GAL::Parser subclasses for sequence alteration formats.
    </li>
  </ul>

  <li>
    <a href="GAL/Storage.html">GAL::Storage</a> - Provides backend storage in an SQLite database.
  </li>

  <ul>
    <li>
      <a href="GAL/Storage/SQLite.html">GAL::Storage::SQLite</a> - Provides backend storage in an SQLite database.
    </li>

    <!--
    <li>
      <a href="GAL/Storage/mysql.html">GAL::Storage::mysql</a> - Provides backend storage in an MySQL database.  Not implimented.
    </li>

    <li>
      <a href="GAL/Storage/RAM.html">GAL::Storage::RAM</a> - Provides backend storage in an RAM database.  Not implimented.
    </li>
    -->

  </ul>

  <li>
    <a href="GAL/Schema.html">GAL::Schema</a> - A class that provides
    basic configuration to DBIx::Class.
  </li>

  <ul>
    <li>
      <a href="GAL/Schema/Result/Feature.html">GAL::Schema::Result::Feature</a>
      - A DBIx::Class subclass that describes the structure of the
      feature table in the database.  The class also impliments methods
      that are inherited by GAL::Schema::Result::Feature subclasses.
    </li>

    <ul>
      <li>
        <a href="GAL/Schema/Result/Feature/sequence_feature.html">GAL::Schema::Result::Feature::sequence_feature</a>
        - A basic sequence_feature class that provides methods for all
        subclasses.  The methods are actually implimented in
        GAL::Schema::Result::Feature, but inherited here so that all
        sequence_feature types for which no subclass has been written will
        inherit all basic sequence_feature behavior from this class.
      </li>

      <ul>
      	<li>
      	  <a href="GAL/Schema/Result/Feature/gene.html">GAL::Schema::Result::Feature::gene</a>
      	  - A class to provide gene specific methods.
      	</li>

      	<li>
      	  <a href="GAL/Schema/Result/Feature/transcript.html">GAL::Schema::Result::Feature::transcript</a>
      	  - A class to provide transcript specific methods.
      	</li>

	<ul>
      	  <li>
      	    <a href="GAL/Schema/Result/Feature/mrna.html">GAL::Schema::Result::Feature::mrna</a>
      	    - A class to provide mRNA specific methods.
      	  </li>
	</ul>

      	<li>
      	  <a href="GAL/Schema/Result/Feature/exon.html">GAL::Schema::Result::Feature::exon</a>
      	  - A class to provide exon specific methods.
      	</li>

      	<li>
      	  <a href="GAL/Schema/Result/Feature/intron.html">GAL::Schema::Result::Feature::intron</a>
      	  - A class to provide intron specific methods.
      	</li>

      	<li>
      	  <a href="GAL/Schema/Result/Feature/cds.html">GAL::Schema::Result::Feature::cds</a>
      	  - A class to provide CDS specific methods.
      	</li>

      	<li>
      	  <a href="GAL/Schema/Result/Feature/protein.html">GAL::Schema::Result::Feature::protein</a>
      	  - A class to provide protein specific methods.
      	</li>

      	<li>
      	  <a href="GAL/Schema/Result/Feature/five_prime_utr.html">GAL::Schema::Result::Feature::five_prime_utr</a>
      	  - A class to provide five_prime_UTR specific methods.
      	</li>

      	<li>
      	  <a href="GAL/Schema/Result/Feature/three_prime_utr.html">GAL::Schema::Result::Feature::three_prime_utr</a>
      	  - A class to provide three_prime_UTR specific methods.
      	</li>

      	<li>
      	  <a href="GAL/Schema/Result/Feature/sequence_alteration.html">GAL::Schema::Result::Feature::sequence_alteration</a>
      	  - A class to provide sequence_alteration specific methods.
      	</li>

      	<li>
      	  <a href="GAL/Schema/Result/Feature/template.html">GAL::Schema::Result::Feature::template</a>
      	  - A template class for use in writing new
      	  GAL::Schema::Result::Feature subclasses.
      	</li>
      </ul>
    </ul>
    <li>
      <a href="GAL/Schema/Result/Attribute.html">GAL::Schema::Result::Attribute</a>
    </li>

    <li>
      <a href="GAL/Schema/Result/Relationship.html">GAL::Schema::Result::Relationship</a>
    </li>
  </ul>

  <li>
    <a href="GAL/Interval.html">GAL::Interval</a> - Classes that
    provide interval (sequence ranges) aggregation and analysis
    functions for GAL.  Most of these are currently in developement
    and thus not may not be functional or stable.
  </li>

  <ul>
    <li>
      <a href="GAL/Interval/Span.html">GAL::Interval::Span</a>
    </li>

    <!--
    <li>
      <a href="GAL/Interval/BitVector.html">GAL::Interval::BitVector</a>
      - Not yet implimented
    </li>

    <li>
      <a href="GAL/Interval/NCL.html">GAL::Interval::NCL</a> - Not yet implimented
    </li>
    -->

  </ul>


  <li>
    <a href="GAL/List.html">GAL::List</a> - List aggregation and analysis methods
  </li>

  <ul>
    <li>
      <a href="GAL/List/Categorical.html">GAL::List::Categorical</a> - List aggregation and analysis functions for categorical lists
    </li>
    
    <li>
      <a href="GAL/List/Numeric.html">GAL::List::Numeric</a> - List aggregation and analysis functions for numerical lists
    </li>
  </ul>

  <li>
    <a href=""></a>
  </li>

  <li>
    <a href=""></a>
  </li>

  <li>
    <a href=""></a>
  </li>

</ul>

</BODY>
</HTML>
