<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en">
<head>
	<title>GAL::Parser::VCFv4_0</title>
<link rel="stylesheet" type="text/css" href="../podstyle.css" /><meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
  <script type="text/javascript">

  function toggleCode( id ) {
    if ( document.getElementById )
      elem = document.getElementById( id );
    else if ( document.all )
      elem = eval( "document.all." + id );
    else
      return false;

    elemStyle = elem.style;

    if ( elemStyle.display != "block" ) {
      elemStyle.display = "block"
    } else {
      elemStyle.display = "none"
    }

    return true;
  }
  document.writeln( "<style type=\"text/css\">div.method-source-code { display: none }</style>" )
  </script></head>
<body>
<div class="box">
  <h1 class="t1">GAL</h1>
  <table>
    <tr>
      <td class="label">Description</td>
      <td class="cell">GAL - The Genome Annotation Library</td>
    </tr>
  </table>
</div>
<div class="path">
  <a href="../index.html">GAL</a> &gt; Perl Modules &gt;
  Parser-VCFv4_0
</div>
<div>
<a href="../src/Parser/VCFv4_0.pm">Source</a>
</div>


<div class="pod">
<!-- INDEX START -->
<h3 id="TOP">Index</h3>
<ul>
	<li><a href="#NAME">NAME</a></li>
	<li><a href="#VERSION">VERSION</a></li>
	<li><a href="#SYNOPSIS">SYNOPSIS</a></li>
	<li><a href="#DESCRIPTION">DESCRIPTION</a></li>
	<li><a href="#METHODS">METHODS</a><br />
<ul>
	<li><a href="#new">new</a></li>
	<li><a href="#parse_record">parse_record</a></li>
</ul>
</li>
	<li><a href="#DIAGNOSTICS">DIAGNOSTICS</a></li>
	<li><a href="#CONFIGURATION_AND_ENVIRONMENT">CONFIGURATION AND ENVIRONMENT</a></li>
	<li><a href="#DEPENDENCIES">DEPENDENCIES</a></li>
	<li><a href="#INCOMPATIBILITIES">INCOMPATIBILITIES</a></li>
	<li><a href="#BUGS_AND_LIMITATIONS">BUGS AND LIMITATIONS</a></li>
	<li><a href="#AUTHOR">AUTHOR</a></li>
	<li><a href="#LICENCE_AND_COPYRIGHT">LICENCE AND COPYRIGHT</a></li>
	<li><a href="#DISCLAIMER_OF_WARRANTY">DISCLAIMER OF WARRANTY</a></li>
</ul>
<hr />
<!-- INDEX END -->

<h1 id="NAME">NAME <a href="#TOP" class="toplink"><img alt="^" src="../up.gif" /></a></h1>

<p>GAL::Parser::VCFv4_0 - &lt;One line description of module's purpose here&gt;</p>

<h1 id="VERSION">VERSION <a href="#TOP" class="toplink"><img alt="^" src="../up.gif" /></a></h1>

<p>This document describes GAL::Parser::VCFv4_0 version 0.2.0</p>

<h1 id="SYNOPSIS">SYNOPSIS <a href="#TOP" class="toplink"><img alt="^" src="../up.gif" /></a></h1>

<pre>     use GAL::Parser::VCFv4_0;

</pre><h1 id="DESCRIPTION">DESCRIPTION <a href="#TOP" class="toplink"><img alt="^" src="../up.gif" /></a></h1>

<h1 id="METHODS">METHODS <a href="#TOP" class="toplink"><img alt="^" src="../up.gif" /></a></h1>

<h2 id="new">new</h2>

<p><a href="#" onclick="toggleCode('method_new');return false;">[Source]</a></p>
                                        <div class="method-source-code" id="method_new">
                                        <pre>

sub new {
	my ($class, @args) = @_;
	my $self = $class->SUPER::new(@args);
	return $self;
}
</pre></div>

<pre>     Title   : new
     Usage   : GAL::Parser::VCFv4_0-&gt;new();
     Function: Creates a GAL::Parser::VCFv4_0 object;
     Returns : A GAL::Parser::VCFv4_0 object
     Args    :

</pre><h2 id="parse_record">parse_record</h2>

<p><a href="#" onclick="toggleCode('method_parse_record');return false;">[Source]</a></p>
                                        <div class="method-source-code" id="method_parse_record">
                                        <pre>

sub parse_record {
    my ($self, $data) = @_;

    ##fileformat=VCFv4.0
    ##INFO=<ID=DP,Number=1,Type=Integer,Description="Total Depth">
    ##INFO=<ID=HM2,Number=0,Type=Flag,Description="HapMap2 membership">
    ##INFO=<ID=HM3,Number=0,Type=Flag,Description="HapMap3 membership">
    ##reference=human_b36_both.fasta
    ##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
    ##FORMAT=<ID=DP,Number=1,Type=Integer,Description="Read Depth">
    ##FORMAT=<ID=CB,Number=1,Type=String,Description="Called by S(Sanger), M(UMich), B(BI)">
    ##rsIDs=dbSNP b129 mapped to NCBI 36.3, August 10, 2009
    ##INFO=<ID=AC,Number=.,Type=Integer,Description="Allele count in genotypes">
    ##INFO=<ID=AN,Number=1,Type=Integer,Description="Total number of alleles in called genotypes">
    #CHROM POS ID REF ALT QUAL FILTER INFO FORMAT SAMPLE1

    my %record;

    # We didn't get $data as a hash because we didn't give field names to the reader.
    # This allows us to get the data as an array and deal with multiple genotypes per line.
    @record{qw(chrom pos id ref alt qual filter info format)} = splice(@{$data}, 0, 9);

    my $seqid      = $record{chrom}; # =~ /^chr/ ? $record{chrom} : 'chr' . $record{chrom};
    my $source;
    my $type       = 'SNV';
    my $start      = $record{pos};
    my $end        = $start;
    my $score      = $record{qual};
    my $strand     = '+',
    my $phase      = '.';

    #Validate var
    my $reference_seq = uc $record{ref};

    my @variant_seqs = split /,/, $record{alt};

    map {$_ = uc $_} @variant_seqs;

    my @all_seqs = ($reference_seq, @variant_seqs);

    # INFO Column Keys
    # VFC 4.0
    # AA ancestral allele
    # AC allele count in genotypes, for each ALT allele, in the same order as listed
    # AF allele frequency for each ALT allele in the same order as listed: use this when estimated from primary data, not called genotypes
    # AN total number of alleles in called genotypes
    # BQ RMS base quality at this position
    # CIGAR cigar string describing how to align an alternate allele to the reference allele
    # DB dbSNP membership
    # DP combined depth across samples, e.g. DP=154
    # END end position of the variant described in this record (esp. for CNVs)
    # H2 membership in hapmap2
    # MQ RMS (root mean square) mapping quality, e.g. MQ=52
    # MQ0 Number of MAPQ == 0 reads covering this record
    # NS Number of samples with data
    # SB strand bias at this position
    # SOMATIC indicates that the record is a somatic mutation, for cancer genomics
    # VALIDATED validated by follow-up experiment

    my %info;
    my @pairs = split /;/, $record{info};
    for my $pair (@pairs) {
      my ($key, $value) = split /=/, $pair;
      $value = defined $value ? $value : '';
      my @values = split /,/, $value;
      push @{$info{$key}}, @values;
    }




    my @format_order = split /:/, $record{format};

    my %individual_data;
    my ($header) = $self->reader->headers;
    my @header_data = split /\s+/, $header;
    splice(@header_data, 0, 9);
    @individual_data{@header_data} = @{$data};

    my @features;
    for my $individual_id (sort keys %individual_data) {

	my @individual_values = split /:/, $individual_data{$individual_id};
	my %individual_hash;
	map {$individual_hash{$_} = shift @individual_values} @format_order;

	my ($var1, $phased, $var2) =
	    split //, $individual_hash{GT};
	my $this_ref = $reference_seq;

	my %variant_seq_hash = map {$_ => 1} @all_seqs[$var1, $var2];
	my @variant_seqs = keys %variant_seq_hash;

	my @these_all_seqs = ($this_ref, @variant_seqs);
	my @lengths = map {length($_)} @these_all_seqs;

	my $this_start = $start;
	my $this_end   = $end;
	#INDEL
	if (grep {$_ != 1} @lengths) {
	  my ($min_length) = sort {$a <=> $b} @lengths;
	  map {substr($_, 0, $min_length, '')} @these_all_seqs;
	  map {$_ ||= '-'} @these_all_seqs;
	  my $start_adjust = length($reference_seq) - length($these_all_seqs[0]);
	  $this_start += $start_adjust;
	  my $end_adjust = length($these_all_seqs[0]) - 1;
	  $this_end = $this_start + $end_adjust;

	  $this_ref = shift @these_all_seqs;
	  @variant_seqs = @these_all_seqs;

	  if ($this_ref eq '-') {
	    $type = 'insertion';
	  }
	  elsif (grep {$_ eq '-'} @variant_seqs) {
	    $type = 'deletion';
	  }
	  else {
	    $type = 'complex_substitution';
	  }
	}

	my $read_count = $individual_hash{DP} || $info{DP}[0] || undef;

	# Make sure that at least one variant sequence differs from
	# the reference sequence.
	next unless grep {$_ ne $this_ref} @variant_seqs;

	# Format Column Keys
	# GT : genotype, encoded as alleles values separated by either of
	# â/â or â|â, e.g. The allele values are 0 for the reference
	# allele (what is in the reference sequence), 1 for the first
	# allele listed in ALT, 2 for the second allele list in ALT and
	# so on. For diploid calls examples could be 0/1 or 1|0 etc. For
	# haploid calls, e.g. on Y, male X, mitochondrion, only one
	# allele value should be given. All samples must have GT call
	# information; if a call cannot be made for a sample at a given
	# locus, â.â must be specified for each missing allele in the GT
	# field (for example ./. for a diploid). The meanings of the
	# separators are:
	#       / : genotype unphased
	#       | : genotype phased
	# DP : read depth at this position for this sample (Integer)
	# FT : sample genotype filter indicating if this genotype was
	# âcalledâ (similar in concept to the FILTER field). Again, use
	# PASS : to indicate that all filters have been passed, a
	# semi-colon separated list of codes for filters that fail, or
	# â.â to indicate that filters have not been applied. These
	# values should be described in the meta-information in the same
	# way as FILTERs (Alphanumeric String)
	# GL : three floating point log10-scaled likelihoods for
	# AA,AB,BB genotypes where A=ref and B=alt; not applicable if
	# site is not biallelic. For example: GT:GL
	# 0/1:-323.03,-99.29,-802.53 (Numeric)
	# GQ : genotype quality, encoded as a phred quality
	# -10log_10p(genotype call is wrong) (Numeric)
	# HQ : haplotype qualities, two phred qualities comma separated
	# (Numeric)

	my $source = $individual_id;
	my $feature_id = join ':', ($seqid, $source, $start);

	my $attributes = {Reference_seq => [$this_ref],
			  Variant_seq   => \@variant_seqs,
			  ID            => [$feature_id],
		      };
	$attributes->{Total_reads} = [$read_count] if $read_count;

	my $feature = {feature_id => $feature_id,
		       seqid      => $seqid,
		       source     => $individual_id,
		       type       => $type,
		       start      => $this_start,
		       end        => $this_end,
		       score      => $score,
		       strand     => $strand,
		       phase      => $phase,
		       attributes => $attributes,
		   };
	push @features, $feature;
    }
    return scalar @features ? \@features : undef;
}
</pre></div>

<pre> Title   : parse_record
 Usage   : $a = $self-&gt;parse_record();
 Function: Parse the data from a record.
 Returns : A hash ref needed by Feature.pm to create a Feature object
 Args    : A hash ref of fields that this sub can understand (In this case GFF3).

</pre><h1 id="DIAGNOSTICS">DIAGNOSTICS <a href="#TOP" class="toplink"><img alt="^" src="../up.gif" /></a></h1>

<dl>
	<dt><code>Error message here, perhaps with %s placeholders</code></dt>
		<dd><p>[Description of error here]</p></dd>
	<dt><code>Another error message here</code></dt>
		<dd><p>[Description of error here]</p></dd>
		<dd><p>[Et cetera, et cetera]</p></dd>
</dl>
<h1 id="CONFIGURATION_AND_ENVIRONMENT">CONFIGURATION AND ENVIRONMENT <a href="#TOP" class="toplink"><img alt="^" src="../up.gif" /></a></h1>

<p>&lt;GAL::Parser::VCFv4_0&gt; requires no configuration files or environment variables.</p>

<h1 id="DEPENDENCIES">DEPENDENCIES <a href="#TOP" class="toplink"><img alt="^" src="../up.gif" /></a></h1>

<p>None.</p>

<h1 id="INCOMPATIBILITIES">INCOMPATIBILITIES <a href="#TOP" class="toplink"><img alt="^" src="../up.gif" /></a></h1>

<p>None reported.</p>

<h1 id="BUGS_AND_LIMITATIONS">BUGS AND LIMITATIONS <a href="#TOP" class="toplink"><img alt="^" src="../up.gif" /></a></h1>

<p>No bugs have been reported.</p>

<p>Please report any bugs or feature requests to:
barry.moore@genetics.utah.edu</p>

<h1 id="AUTHOR">AUTHOR <a href="#TOP" class="toplink"><img alt="^" src="../up.gif" /></a></h1>

<p>Barry Moore &lt;barry.moore@genetics.utah.edu&gt;</p>

<h1 id="LICENCE_AND_COPYRIGHT">LICENCE AND COPYRIGHT <a href="#TOP" class="toplink"><img alt="^" src="../up.gif" /></a></h1>

<p>Copyright (c) 2012, Barry Moore &lt;barry.moore@genetics.utah.edu&gt;.  All rights reserved.</p>

<pre>    This module is free software; you can redistribute it and/or
    modify it under the same terms as Perl itself (See LICENSE).

</pre><h1 id="DISCLAIMER_OF_WARRANTY">DISCLAIMER OF WARRANTY <a href="#TOP" class="toplink"><img alt="^" src="../up.gif" /></a></h1>

<p>BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE &quot;AS IS&quot; WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.</p>

<p>IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.</p>


</div><div class="footer">generated by <a href="http://search.cpan.org/perldoc?Pod/ProjectDocs">Pod::ProjectDocs</a></div></body>
</html>
