<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en">
<head>
	<title>GAL::Schema::Result::Feature::transcript</title>
<link rel="stylesheet" type="text/css" href="../../../podstyle.css" /><meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
  <script type="text/javascript">

  function toggleCode( id ) {
    if ( document.getElementById )
      elem = document.getElementById( id );
    else if ( document.all )
      elem = eval( "document.all." + id );
    else
      return false;

    elemStyle = elem.style;

    if ( elemStyle.display != "block" ) {
      elemStyle.display = "block"
    } else {
      elemStyle.display = "none"
    }

    return true;
  }
  document.writeln( "<style type=\"text/css\">div.method-source-code { display: none }</style>" )
  </script></head>
<body>
<div class="box">
  <h1 class="t1">GAL</h1>
  <table>
    <tr>
      <td class="label">Description</td>
      <td class="cell">GAL - The Genome Annotation Library</td>
    </tr>
  </table>
</div>
<div class="path">
  <a href="../../../index.html">GAL</a> &gt; Perl Modules &gt;
  Schema-Result-Feature-transcript
</div>
<div>
<a href="../../../src/Schema/Result/Feature/transcript.pm">Source</a>
</div>


<div class="pod">
<!-- INDEX START -->
<h3 id="TOP">Index</h3>
<ul>
	<li><a href="#NAME">NAME</a></li>
	<li><a href="#VERSION">VERSION</a></li>
	<li><a href="#SYNOPSIS">SYNOPSIS</a></li>
	<li><a href="#DESCRIPTION">DESCRIPTION</a></li>
	<li><a href="#METHODS">METHODS</a><br />
<ul>
	<li><a href="#exons">exons</a></li>
	<li><a href="#introns">introns</a></li>
	<li><a href="#infer_introns">infer_introns</a></li>
	<li><a href="#mature_seq_genomic">mature_seq_genomic</a></li>
	<li><a href="#mature_seq">mature_seq</a></li>
	<li><a href="#length">length</a></li>
	<li><a href="#coordinate_map">coordinate_map</a></li>
	<li><a href="#genome2me">genome2me</a></li>
	<li><a href="#me2genome">me2genome</a></li>
	<li><a href="#AED">AED</a></li>
</ul>
</li>
	<li><a href="#DIAGNOSTICS">DIAGNOSTICS</a></li>
	<li><a href="#CONFIGURATION_AND_ENVIRONMENT">CONFIGURATION AND ENVIRONMENT</a></li>
	<li><a href="#DEPENDENCIES">DEPENDENCIES</a></li>
	<li><a href="#INCOMPATIBILITIES">INCOMPATIBILITIES</a></li>
	<li><a href="#BUGS_AND_LIMITATIONS">BUGS AND LIMITATIONS</a></li>
	<li><a href="#AUTHOR">AUTHOR</a></li>
	<li><a href="#LICENCE_AND_COPYRIGHT">LICENCE AND COPYRIGHT</a></li>
	<li><a href="#DISCLAIMER_OF_WARRANTY">DISCLAIMER OF WARRANTY</a></li>
</ul>
<hr />
<!-- INDEX END -->

<h1 id="NAME">NAME <a href="#TOP" class="toplink"><img alt="^" src="../../../up.gif" /></a></h1>

<p>GAL::Schema::Result::Feature::transcript - A transcript object for the GAL
Library</p>

<h1 id="VERSION">VERSION <a href="#TOP" class="toplink"><img alt="^" src="../../../up.gif" /></a></h1>

<p>This document describes GAL::Schema::Result::Feature::transcript
version 0.2.0</p>

<h1 id="SYNOPSIS">SYNOPSIS <a href="#TOP" class="toplink"><img alt="^" src="../../../up.gif" /></a></h1>

<pre>    use GAL::Annotation;
    my $feat_store = GAL::Annotation-&gt;new(storage =&gt; $feat_store_args,
					  parser  =&gt; $parser_args,
					  fasta   =&gt; $fasta_args,
					 );

    $feat_store-&gt;load_files(files =&gt; $feature_file,
			    mode  =&gt; 'overwrite',
			    );

    my $features = $feat_store-&gt;schema-&gt;resultset('Feature');

    my $genes = $features-&gt;search({type =&gt; 'gene'});
    while (my $gene = $genes-&gt;next) {
      my $transcripts = $gene-&gt;transcripts;
      while (my $transcript = $transcripts-&gt;next) {
	my $id    = $transcript-&gt;feature_id;
	my $start = $transcript-&gt;start;
	my $end   = $transcript-&gt;end;
      }

    }

</pre><h1 id="DESCRIPTION">DESCRIPTION <a href="#TOP" class="toplink"><img alt="^" src="../../../up.gif" /></a></h1>

<p>&lt;GAL::Schema::Result::Feature::transcript&gt; provides a &lt;GAL::Schema::Result::Feature&gt;
subclass for transcript specific behavior.</p>

<h1 id="METHODS">METHODS <a href="#TOP" class="toplink"><img alt="^" src="../../../up.gif" /></a></h1>

<h2 id="exons">exons</h2>

<p><a href="#" onclick="toggleCode('method_exons');return false;">[Source]</a></p>
                                        <div class="method-source-code" id="method_exons">
                                        <pre>

sub exons {
  my $self = shift;

  if (wantarray) {
      return grep {$_->type eq 'exon'} $self->children->all;
  }
  else {
      my $sort_order = ($self->strand eq '-' ?
			{'-desc' => 'end'}   :
			{'-asc'  => 'start'});

      my $exons = $self->children({type => 'exon'},
				  {order_by => $sort_order,
				   distinct => 1});
      return $exons;
  }
}
</pre></div>

<pre> Title   : exons
 Usage   : $exons = $self-&gt;exons
 Function: Get the transcript's exons.
 Returns : In list context returns an unsorted array of exon objects.
	   In scalar context returns an iterator (DBIC::ResultSet)
	   with exons sorted in 5'-&gt;3' on the transcript's strand.
	   Calling in list context is much faster but loads all exons
	   immediately into memory.
 Args    : None

</pre><h2 id="introns">introns</h2>

<p><a href="#" onclick="toggleCode('method_introns');return false;">[Source]</a></p>
                                        <div class="method-source-code" id="method_introns">
                                        <pre>

sub introns {
  my $self = shift;

  if (wantarray) {
      my @introns = grep {$_->type eq 'intron'} $self->children->all;
      $self->infer_introns() unless ref $introns[0];
      return grep {$_->type eq 'intron'} $self->children->all;
  }
  else {
      my $sort_order = ($self->strand eq '-' ?
			{'-desc' => 'end'}   :
			{'-asc'  => 'start'});

      my $introns = $self->children({type => 'intron'},
				  {order_by => $sort_order,
				   distinct => 1});
      if (! $introns->count) {
	$self->infer_introns($introns);
      }

      return $introns;
  }
}
</pre></div>

<pre> Title   : introns
 Usage   : $introns = $self-&gt;introns
 Function: Get the transcript's introns.
 Returns : In list context returns an unsorted array of intron objects.
	   In scalar context returns an iterator (DBIC::ResultSet)
	   with introns sorted in 5'-&gt;3' on the transcript's strand.
	   Calling in list context is much faster but loads all exons
	   immediately into memory.
 Args    : None

</pre><h2 id="infer_introns">infer_introns</h2>

<p><a href="#" onclick="toggleCode('method_infer_introns');return false;">[Source]</a></p>
                                        <div class="method-source-code" id="method_infer_introns">
                                        <pre>

sub infer_introns {
  my ($self, $introns) = @_;

  if (! $introns) {
    my $sort_order;
    if ($self->strand eq '-') {
      $sort_order = {'-desc' => 'end'};
    }
    else {
      $sort_order = {'-asc' => 'start'};
    }

    $introns ||= $self->children({type => 'intron'},
				 {order_by => $sort_order,
				  distinct => 1});
  }

  # Keep the order_by like this so we stay in genomic order
  my @exons = $self->children({type => 'exon'},
			      {order_by => { -asc => 'start' },
			       distinct => 1})->all;

  my @coordinates;
  for my $exon (@exons) {
    push @coordinates, ($exon->start, $exon->end);
  }
  @coordinates = sort {$a <=> $b} @coordinates;

  shift @coordinates;
  pop   @coordinates;

  my %template = (seqid  => $self->seqid,
		  source => $self->source,
		  type   => 'intron',
		  score  => '.',
		  strand => $self->strand,
		  phase  => '.',
		 );

  my $parent_id = $self->feature_id;

  my $count = 1;
  my @introns;
  while (@coordinates) {
    my ($start, $end) = (shift @coordinates,
			 shift @coordinates);
    my $feature_id = $parent_id . ':intron:';
    $feature_id .= sprintf("%03d", $count++);
    my @attrbs = ({att_key    => 'ID',
		   att_value  => $feature_id},
		  {att_key    => 'Parent',
		   att_value  => $parent_id},
		 );
    my @rels = {parent => $parent_id,
		child  => $feature_id};

    my %intron = %template;
    @intron{qw(feature_id start end attributes my_parents)} =
      ($feature_id, $start, $end, \@attrbs, \@rels);
    my $intron = $introns->find_or_create(\%intron);
    bless $intron, 'GAL::Schema::Result::Feature::intron';
    push @introns, $intron;
  }
  $introns->set_cache(\@introns);
}
</pre></div>

<pre> Title   : infer_introns
 Usage   : $introns = $self-&gt;infer_introns
 Function: Infer introns for the transcript.
 Returns : A DBIx::Class::Result object loaded up with introns
 Args    : None

</pre><h2 id="mature_seq_genomic">mature_seq_genomic</h2>

<p><a href="#" onclick="toggleCode('method_mature_seq_genomic');return false;">[Source]</a></p>
                                        <div class="method-source-code" id="method_mature_seq_genomic">
                                        <pre>

sub mature_seq_genomic {
  my $self = shift;

  my $mature_seq_genomic;
  map {$mature_seq_genomic .= $_->genomic_seq}
    sort {$a->start <=> $b->start }$self->exons;
 #  my $exons = $self->exons;
 # EXON:
 #  while (my $exon = $exons->next) {
 #      my $seq = $exon->seq;
 #      if (! $seq) {
 # 	  $self->warn('no_sequence_available_for_feature',
 # 		      $exon->feature_id);
 # 	  next EXON;
 #      }
 #      $mature_seq_genomic .= $seq;
 #  }
  return $mature_seq_genomic;
}
</pre></div>

<pre> Title   : mature_seq_genomic
 Usage   : $seq = $self-&gt;mature_seq_genomic
 Function: Get the transcripts spliced genomic sequence (not reverse
	   complimented for minus strand features).
 Returns : A text string of the transcripts splice genomic sequence.
 Args    : None

</pre><h2 id="mature_seq">mature_seq</h2>

<p><a href="#" onclick="toggleCode('method_mature_seq');return false;">[Source]</a></p>
                                        <div class="method-source-code" id="method_mature_seq">
                                        <pre>

sub mature_seq {
  my $self = shift;

  my $mature_seq = $self->mature_seq_genomic;
  $mature_seq = $self->annotation->revcomp($mature_seq)
    if ($self->strand eq '-');
  return $mature_seq;
}
</pre></div>

<pre> Title   : mature_seq
 Usage   : $seq = $self-&gt;mature_seq
 Function: Get the transcripts spliced sequence reverse complimented for minus
	   strand features.
 Returns : A text string of the transcripts splice genomic sequence.
 Args    : None

</pre><h2 id="length">length</h2>

<p><a href="#" onclick="toggleCode('method_length');return false;">[Source]</a></p>
                                        <div class="method-source-code" id="method_length">
                                        <pre>

sub length {
  my $self = shift;
  my $length;
  map {$length += $_->length} $self->exons;
  $self->warn('transcript_has_no_exons', $self->feature_id) unless $length;
  return $length;
}
</pre></div>

<pre> Title   : length
 Usage   : $length = $self-&gt;length
 Function: Get the length of the mature transcript - the sum of all of its
	   exon lengths.
 Returns : An integer
 Args    : None

</pre><h2 id="coordinate_map">coordinate_map</h2>

<p><a href="#" onclick="toggleCode('method_coordinate_map');return false;">[Source]</a></p>
                                        <div class="method-source-code" id="method_coordinate_map">
                                        <pre>

sub coordinate_map {
  my $self = shift;
  if (! $self->{coordinate_map}) {
    my $strand = $self->strand;
    my $length = $self->length;
    my %coordinate_map;
    my @exons = $self->exons;
    my ($transcript_position, $increment);
    if ($strand eq '-') {
      $transcript_position = $length - 1;
      $increment = -1;
    }
    else {
      $transcript_position = 1;
      $increment = 1;
    }
    for my $exon (@exons) {
      my $start = $exon->start;
      my $end   = $exon->end;
      for my $genomic_position ($start .. $end) {
	$coordinate_map{genome2me}{$genomic_position}    = $transcript_position;
	$coordinate_map{me2genome}{$transcript_position} = $genomic_position;
	$transcript_position += $increment;
      }
    }
    $self->{coordinate_map} = \%coordinate_map;
  }
  return wantarray ? %{$self->{coordinate_map}} : $self->{coordinate_map};
}
</pre></div>

<pre> Title   : coordinate_map
 Usage   : $map = $self-&gt;coordinate_map
 Function: Get the coordinate map the which has the structure:
	   $coordinate_map{genome2me}{$genomic_position}    = $transcript_position;
	   $coordinate_map{me2genome}{$transcript_position} = $genomic_position;
	   And thus can be used to map feature to genomic coordinates and vice versa.
 Returns : A hash or reference of the above map.
 Args    : None

</pre><h2 id="genome2me">genome2me</h2>

<p><a href="#" onclick="toggleCode('method_genome2me');return false;">[Source]</a></p>
                                        <div class="method-source-code" id="method_genome2me">
                                        <pre>

sub genome2me {
  my ($self, @coordinates) = @_;
  my $coordinate_map = $self->coordinate_map;
  my @my_coordinates;
  for my $coordinate (@coordinates) {
    push @my_coordinates, $coordinate_map->{genome2me}{$coordinate};
  }
  return wantarray ? @my_coordinates : \@my_coordinates;
}
</pre></div>

<pre> Title   : genome2me
 Usage   : $my_coordinates = $self-&gt;genome2me(@genome_coordinates);
 Function: Transform genomic coordinates to mature transcript coordinates.
 Returns : An array or reference of mature transcript coordinates.
 Args    : An array reference of genomic coordinates.

</pre><h2 id="me2genome">me2genome</h2>

<p><a href="#" onclick="toggleCode('method_me2genome');return false;">[Source]</a></p>
                                        <div class="method-source-code" id="method_me2genome">
                                        <pre>

sub me2genome {
  my ($self, @coordinates) = @_;
  my $coordinate_map = $self->coordinate_map;
  my @genomic_coordinates;
  for my $coordinate (@coordinates) {
    push @genomic_coordinates, $coordinate_map->{me2genome}{$coordinate};
  }
  return wantarray ? @genomic_coordinates : \@genomic_coordinates;
}
</pre></div>

<pre> Title   : me2genome
 Usage   : $genomic_coordinates = $self-&gt;me2genome(@my_coordinates);
 Function: Transform mature transcript coordinates to genomic coordinates.
 Returns : An array or reference of genomic coordinates.
 Args    : An array reference of mature transcript coordinates.

</pre><h2 id="AED">AED</h2>

<p><a href="#" onclick="toggleCode('method_AED');return false;">[Source]</a></p>
                                        <div class="method-source-code" id="method_AED">
                                        <pre>

sub AED {

  my ($transcriptA, $transcriptB) = @_;

  my @exonsA = $transcriptA->exons;
  my $setA = Set::IntSpan::Fast->new;
  map {$setA->add_range($_->start, $_->end)} @exonsA;
  my $i = scalar $setA->as_array;

  my @exonsB = $transcriptB->exons;
  my $setB = Set::IntSpan::Fast->new;
  map {$setB->add_range($_->start, $_->end)} @exonsB;
  my $j = scalar $setB->as_array;

  my $int = scalar $setA->intersection($setB)->as_array;

  my $sn  = $int/$j;
  my $sp  = $int/$i;
  my $ac  = ($sn + $sp)/2;
  my $aed = 1 - $ac;

  return $aed;
}
</pre></div>

<pre> Title   : AED
 Usage   : $aed = $self-&gt;AED($transcript);
 Function: Calculate the annotation edit distance (AED) between two
	   transcripts. (PMID:19236712)
 Returns : An array or reference of genomic coordinates.
 Args    : A GAL::Schema::Result::Feature::transcript (or is_a) object

</pre><h1 id="DIAGNOSTICS">DIAGNOSTICS <a href="#TOP" class="toplink"><img alt="^" src="../../../up.gif" /></a></h1>

<p>&lt;GAL::Schema::Result::Feature::transcript&gt; does not throw any warnings
or error messages.</p>

<h1 id="CONFIGURATION_AND_ENVIRONMENT">CONFIGURATION AND ENVIRONMENT <a href="#TOP" class="toplink"><img alt="^" src="../../../up.gif" /></a></h1>

<p>&lt;GAL::Schema::Result::Feature::transcript&gt; requires no configuration
files or environment variables.</p>

<h1 id="DEPENDENCIES">DEPENDENCIES <a href="#TOP" class="toplink"><img alt="^" src="../../../up.gif" /></a></h1>

<p>&lt;GAL::Schema::Result::Feature::sequence_feature&gt;</p>

<h1 id="INCOMPATIBILITIES">INCOMPATIBILITIES <a href="#TOP" class="toplink"><img alt="^" src="../../../up.gif" /></a></h1>

<p>None reported.</p>

<h1 id="BUGS_AND_LIMITATIONS">BUGS AND LIMITATIONS <a href="#TOP" class="toplink"><img alt="^" src="../../../up.gif" /></a></h1>

<p>No bugs have been reported.</p>

<p>Please report any bugs or feature requests to:
barry.moore@genetics.utah.edu</p>

<h1 id="AUTHOR">AUTHOR <a href="#TOP" class="toplink"><img alt="^" src="../../../up.gif" /></a></h1>

<p>Barry Moore &lt;barry.moore@genetics.utah.edu&gt;</p>

<h1 id="LICENCE_AND_COPYRIGHT">LICENCE AND COPYRIGHT <a href="#TOP" class="toplink"><img alt="^" src="../../../up.gif" /></a></h1>

<p>Copyright (c) 2012, Barry Moore &lt;barry.moore@genetics.utah.edu&gt;.  All rights reserved.</p>

<pre>    This module is free software; you can redistribute it and/or
    modify it under the same terms as Perl itself (See LICENSE).

</pre><h1 id="DISCLAIMER_OF_WARRANTY">DISCLAIMER OF WARRANTY <a href="#TOP" class="toplink"><img alt="^" src="../../../up.gif" /></a></h1>

<p>BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE &quot;AS IS&quot; WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.</p>

<p>IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.</p>


</div><div class="footer">generated by <a href="http://search.cpan.org/perldoc?Pod/ProjectDocs">Pod::ProjectDocs</a></div></body>
</html>
