<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
   "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
<title>GAL</title>
<link rel="stylesheet" type="text/css" href="podstyle.css" />
<script type="text/javascript">
//<![CDATA[
var managers = [{"records":[{"name":"Annotation","title":"<a href=\"http://search.cpan.org/perldoc?GAL/Annotation\">GAL::Annotation</a> - The Genome Annotation Library","path":"Annotation.pm.html"},{"name":"Base","title":"<a href=\"http://search.cpan.org/perldoc?GAL/Base\">GAL::Base</a> - Base class for the Genome Annotation Library","path":"Base.pm.html"},{"name":"Index","title":"&lt;One line description of module's purpose here&gt;","path":"Index.pm.html"},{"name":"Interval","title":"Interval aggregation and analysis functions for GAL","path":"Interval.pm.html"},{"name":"Interval-BitVector","title":"Provide functions for intervals on sequence","path":"Interval/BitVector.pm.html"},{"name":"Interval-NCL","title":"Provide functions for intervals on sequence","path":"Interval/NCL.pm.html"},{"name":"Interval-Span","title":"Provide functions for intervals on sequence","path":"Interval/Span.pm.html"},{"name":"List","title":"List aggregation and analysis functions for GAL","path":"List.pm.html"},{"name":"List-Categorical","title":"Provide functions for categorical lists","path":"List/Categorical.pm.html"},{"name":"List-Numeric","title":"Provide functions for lists of numbers","path":"List/Numeric.pm.html"},{"name":"Manual","title":"","path":"Manual.pm.html"},{"name":"Parser","title":"Parser objects for the Genome Annotation Library","path":"Parser.pm.html"},{"name":"Parser-23andMe","title":"Parse 23andMe genotype files","path":"Parser/23andMe.pm.html"},{"name":"Parser-VCFv4_0","title":"&lt;One line description of module's purpose here&gt;","path":"Parser/VCFv4_0.pm.html"},{"name":"Parser-basic_snp","title":"Parse SNP files with basic information.","path":"Parser/basic_snp.pm.html"},{"name":"Parser-bed","title":"Parse BED files","path":"Parser/bed.pm.html"},{"name":"Parser-cgi_allJunctionsBeta","title":"GAL::Parser::cgi_allJunctionsBeta - Parse Complete Genomics\ncnvSegmentsDiploidBeta files","path":"Parser/cgi_allJunctionsBeta.pm.html"},{"name":"Parser-cgi_allSvEventsBeta","title":"GAL::Parser::cgi_allSvEventsBeta - Parse Complete Genomics\ncnvSegmentsDiploidBeta files","path":"Parser/cgi_allSvEventsBeta.pm.html"},{"name":"Parser-cgi_cnvSegmentsDiploidBeta","title":"GAL::Parser::cgi_cnvSegmentsDiploidBeta - Parse Complete Genomics\ncnvSegmentsDiploidBeta files","path":"Parser/cgi_cnvSegmentsDiploidBeta.pm.html"},{"name":"Parser-cgi_compact","title":"Parse Complete Genomics compact files","path":"Parser/cgi_compact.pm.html"},{"name":"Parser-cgi_complete","title":"GAL::Parser::cgi_complete - Parse some types of Complete\nGenomics files.","path":"Parser/cgi_complete.pm.html"},{"name":"Parser-cgi_mobileElementInsertionsBeta","title":"GAL::Parser::cgi_mobileElementInsertionsBeta - Parse Complete Genomics\ncnvSegmentsDiploidBeta files","path":"Parser/cgi_mobileElementInsertionsBeta.pm.html"},{"name":"Parser-chinese_indel","title":"GAL::Parser::chinese_indel - Parse indel files from the first Chinese\ngenome","path":"Parser/chinese_indel.pm.html"},{"name":"Parser-chinese_snp","title":"GAL::Parser::chinese_snp - Parse SNV files from the first Chinese\ngenome","path":"Parser/chinese_snp.pm.html"},{"name":"Parser-dbsnp_flat","title":"&lt;One line description of module's purpose here&gt;","path":"Parser/dbsnp_flat.pm.html"},{"name":"Parser-evan_snp","title":"Parse EVAN_SNP files","path":"Parser/evan_snp.pm.html"},{"name":"Parser-gff3","title":"Parse GFF3 files","path":"Parser/gff3.pm.html"},{"name":"Parser-gff3_fast","title":"Parse GFF3_FAST files","path":"Parser/gff3_fast.pm.html"},{"name":"Parser-hapmap_genotypes","title":"&lt;One line description of module's purpose here&gt;","path":"Parser/hapmap_genotypes.pm.html"},{"name":"Parser-hgmd_indel","title":"Parse HGMD_INDEL files","path":"Parser/hgmd_indel.pm.html"},{"name":"Parser-hgmd_snv","title":"Parse HGMD_SNV files","path":"Parser/hgmd_snv.pm.html"},{"name":"Parser-korean_indel","title":"Parse Illumina indel variant files","path":"Parser/korean_indel.pm.html"},{"name":"Parser-korean_snp","title":"Parse SNV files from the first Korean Genome","path":"Parser/korean_snp.pm.html"},{"name":"Parser-maq_cns2snp","title":"Parse MAQ cns2snp files","path":"Parser/maq_cns2snp.pm.html"},{"name":"Parser-na18507_sanger_indel","title":"GAL::Parser::na18507_sanger_indel - Parse Sanger Center indel files\nfor NA18507","path":"Parser/na18507_sanger_indel.pm.html"},{"name":"Parser-ncbi_blast_tab","title":"Parse NCBI_BLAST_TAB files","path":"Parser/ncbi_blast_tab.pm.html"},{"name":"Parser-nhgri_data","title":"Parse NHGRI_DATA files","path":"Parser/nhgri_data.pm.html"},{"name":"Parser-nhgri_data2","title":"Parse NHGRI_DATA2 files","path":"Parser/nhgri_data2.pm.html"},{"name":"Parser-quake_snp","title":"Parse SNP files from Helicos' Stephen Quake genome","path":"Parser/quake_snp.pm.html"},{"name":"Parser-samtools_pileup","title":"Parse SAMTOOLS pileup files","path":"Parser/samtools_pileup.pm.html"},{"name":"Parser-soap_snp","title":"Parse SOAP_SNP files","path":"Parser/soap_snp.pm.html"},{"name":"Parser-solid_snp","title":"Parse SOLiD SNP files","path":"Parser/solid_snp.pm.html"},{"name":"Parser-template","title":"Parse TEMPLATE files","path":"Parser/template.pm.html"},{"name":"Parser-template_sequence_alteration","title":"Parse TEMPLATE files","path":"Parser/template_sequence_alteration.pm.html"},{"name":"Parser-trait_o_matic","title":"Parse TRAIT-O-MATIC SNP files","path":"Parser/trait_o_matic.pm.html"},{"name":"Parser-ucsc_dbsnp_table","title":"Parse UCSC_DBSNP_TABLE files","path":"Parser/ucsc_dbsnp_table.pm.html"},{"name":"Parser-ucsc_gene_table","title":"Parse UCSC gene table files.","path":"Parser/ucsc_gene_table.pm.html"},{"name":"Parser-ucsc_kg_table","title":"Parse TEMPLATE files","path":"Parser/ucsc_kg_table.pm.html"},{"name":"Parser-vaast_smp","title":"Parse VAAST_SMP files","path":"Parser/vaast_smp.pm.html"},{"name":"Parser-venter_indel","title":"GAL::Parser::venter_indel - Parse indel file from the Craig Venter\ngenome","path":"Parser/venter_indel.pm.html"},{"name":"Parser-venter_snp","title":"Parse SNP files from the Craig Venter genome","path":"Parser/venter_snp.pm.html"},{"name":"Parser-wang_xlinked_data","title":"Parse WANG_XLINKED_DATA files","path":"Parser/wang_xlinked_data.pm.html"},{"name":"Parser-watson_cshl","title":"Parse SNP files from James Watson's genome (CSHL version)","path":"Parser/watson_cshl.pm.html"},{"name":"Parser-watson_novel_het","title":"Parse SNP files from James Watson's genome (CSHL version)","path":"Parser/watson_novel_het.pm.html"},{"name":"Parser-watson_novel_hom","title":"Parse SNP files from James Watson's genome (CSHL version)","path":"Parser/watson_novel_hom.pm.html"},{"name":"Reader","title":"Reader objects for the Genome Annotation Library","path":"Reader.pm.html"},{"name":"Reader-DelimitedLine","title":" Delimited file parsing for GAL","path":"Reader/DelimitedLine.pm.html"},{"name":"Reader-RecordParser","title":"Record Parsing using Text::RecordParser","path":"Reader/RecordParser.pm.html"},{"name":"Reader-Template","title":"&lt;One line description of module's purpose here&gt;","path":"Reader/Template.pm.html"},{"name":"Run","title":"&lt;One line description of module's purpose here&gt;","path":"Run.pm.html"},{"name":"Schema","title":"DBIx::Class functionality for the Genome Annotation Library","path":"Schema.pm.html"},{"name":"Schema-Result-Attribute","title":"Access to feature attributes for GAL::Schema::Result::Features","path":"Schema/Result/Attribute.pm.html"},{"name":"Schema-Result-Feature","title":"Base class for all sequence features","path":"Schema/Result/Feature.pm.html"},{"name":"Schema-Result-Feature-cds","title":"A CDS object for the GAL Library","path":"Schema/Result/Feature/cds.pm.html"},{"name":"Schema-Result-Feature-exon","title":" A exon object for the GAL Library","path":"Schema/Result/Feature/exon.pm.html"},{"name":"Schema-Result-Feature-five_prime_utr","title":"GAL::Schema::Result::Feature::five_prime_utr - A five_prime_UTR object\nfor the GAL Library","path":"Schema/Result/Feature/five_prime_utr.pm.html"},{"name":"Schema-Result-Feature-gene","title":" A gene object for the GAL Library","path":"Schema/Result/Feature/gene.pm.html"},{"name":"Schema-Result-Feature-intron","title":" A intron object for the GAL Library","path":"Schema/Result/Feature/intron.pm.html"},{"name":"Schema-Result-Feature-mrna","title":"GAL::Schema::Result::Feature::mrna - &lt;One line description of module's\npurpose here&gt;","path":"Schema/Result/Feature/mrna.pm.html"},{"name":"Schema-Result-Feature-protein","title":"A protein object for the GAL Library","path":"Schema/Result/Feature/protein.pm.html"},{"name":"Schema-Result-Feature-sequence_alteration","title":"GAL::Schema::Result::Feature::sequence_alteration - A\nsequence_alteration object for the GAL library","path":"Schema/Result/Feature/sequence_alteration.pm.html"},{"name":"Schema-Result-Feature-sequence_feature","title":"&lt;One line description of module's purpose here&gt;","path":"Schema/Result/Feature/sequence_feature.pm.html"},{"name":"Schema-Result-Feature-template","title":"A template object for the GAL Library","path":"Schema/Result/Feature/template.pm.html"},{"name":"Schema-Result-Feature-three_prime_utr","title":"GAL::Schema::Result::Feature::three_prime_utr - A three_prime_UTR\nobject for the GAL Library","path":"Schema/Result/Feature/three_prime_utr.pm.html"},{"name":"Schema-Result-Feature-transcript","title":"GAL::Schema::Result::Feature::transcript - A transcript object for the GAL\nLibrary","path":"Schema/Result/Feature/transcript.pm.html"},{"name":"Schema-Result-Relationship","title":"Access to relationships for &lt;GAL::Schema::Result::Feature&gt; objects.","path":"Schema/Result/Relationship.pm.html"},{"name":"Schema-ResultSet-Feature","title":"","path":"Schema/ResultSet/Feature.pm.html"},{"name":"SchemaAnnotation","title":"","path":"SchemaAnnotation.pm.html"},{"name":"Storage","title":"Base class for GAL storage engines","path":"Storage.pm.html"},{"name":"Storage-RAM","title":"RAM feature storage for GAL","path":"Storage/RAM.pm.html"},{"name":"Storage-SQLite","title":"SQLite feature storage for GAL","path":"Storage/SQLite.pm.html"},{"name":"Storage-mysql","title":"MySQL feature storage for GAL","path":"Storage/mysql.pm.html"}],"desc":"Perl Modules"},{"records":[{"name":"t-devel_scripts-annotation_basic","title":"","path":"t/devel_scripts/annotation_basic.pl.html"},{"name":"t-devel_scripts-annotation_existing","title":"","path":"t/devel_scripts/annotation_existing.pl.html"},{"name":"t-devel_scripts-annotation_load_file","title":"","path":"t/devel_scripts/annotation_load_file.pl.html"},{"name":"t-devel_scripts-parser_1000_genomes_genotypes","title":"","path":"t/devel_scripts/parser_1000_genomes_genotypes.pl.html"},{"name":"t-devel_scripts-parser_complete_genomics","title":"","path":"t/devel_scripts/parser_complete_genomics.pl.html"},{"name":"t-devel_scripts-parser_dbsnp_flat","title":"","path":"t/devel_scripts/parser_dbsnp_flat.pl.html"},{"name":"t-devel_scripts-parser_illumina_indel","title":"","path":"t/devel_scripts/parser_illumina_indel.pl.html"},{"name":"t-devel_scripts-parser_na18507_indel","title":"","path":"t/devel_scripts/parser_na18507_indel.pl.html"},{"name":"t-devel_scripts-parser_quake","title":"","path":"t/devel_scripts/parser_quake.pl.html"},{"name":"t-devel_scripts-parser_samtools_pileup","title":"","path":"t/devel_scripts/parser_samtools_pileup.pl.html"},{"name":"t-devel_scripts-parser_smith_ant_snps","title":"","path":"t/devel_scripts/parser_smith_ant_snps.pl.html"},{"name":"t-devel_scripts-parser_soap_indel","title":"","path":"t/devel_scripts/parser_soap_indel.pl.html"},{"name":"t-devel_scripts-parser_soap_snp","title":"","path":"t/devel_scripts/parser_soap_snp.pl.html"},{"name":"t-devel_scripts-parser_trait_o_matic","title":"","path":"t/devel_scripts/parser_trait_o_matic.pl.html"},{"name":"t-devel_scripts-parser_venter_indel","title":"","path":"t/devel_scripts/parser_venter_indel.pl.html"},{"name":"t-devel_scripts-schema_cursor","title":"","path":"t/devel_scripts/schema_cursor.pl.html"},{"name":"t-devel_scripts-schema_dbi","title":"","path":"t/devel_scripts/schema_dbi.pl.html"},{"name":"t-devel_scripts-schema_hash","title":"","path":"t/devel_scripts/schema_hash.pl.html"},{"name":"t-devel_scripts-schema_mysql","title":"","path":"t/devel_scripts/schema_mysql.pl.html"},{"name":"t-devel_scripts-schema_object","title":"","path":"t/devel_scripts/schema_object.pl.html"},{"name":"t-devel_scripts-storage_mysql","title":"","path":"t/devel_scripts/storage_mysql.pl.html"}],"desc":"Trigger Scripts"}];
function render(pattern) {
    var html = '';
    for ( var i = 0; i < managers.length; i++ ) {
        var manager   = managers[i];
        var rows_html = get_rows_html(manager, pattern);
        var listbox = "<div class='box'><h2 class='t2'>"
                    + manager.desc
                    + "</h2><table width='100%'>"
                    + rows_html
                    + "</table></div>";
        html += listbox;
    }
    var list = document.getElementById('list');
    list.innerHTML = html;
}
function get_rows_html (manager, pattern) {
    var html   = '';
    var regexp = new RegExp( "(" + pattern + ")", "gi");
    var seq    = 0;
    for (var i = 0; i < manager.records.length; i++) {
        var record = manager.records[i];
        if ( record.name.match(regexp) ) {
            var module_name = manager.records[i].name;
            if(pattern != '' ) {
                var replace_text = "<span class='search_highlight'>$1</span>";
                module_name = module_name.replace(regexp, replace_text);
            }
            html += get_record_html(record, module_name, seq);
            seq++;
        }
    }
    return html;
}
function get_record_html (record, name, i) {
    var row_class  = ( i % 2 == 0 ) ? 'r' : 's';
    var row_html   = "<tr class='"
                   + row_class
                   + "'><td nowrap='nowrap'><a href='"
                   + record.path
                   + "'>"
                   + name + "</a></td><td width='99%'><small>"
                   + record.title
                   + "</small></td></tr>";
    return row_html;
}
//]]>
</script>
</head>
<body onload="render('')">
<div class="box">
  <h1 class="t1">GAL</h1>
  <table>
    <tr>
      <td class="label">Description</td>
      <td class="cell">GAL - The Genome Annotation Library</td>
    </tr>
  </table>
</div>

<div class="box">
  <h2 class="t2">Search</h2>
  <table>
    <tr>
       <td class="cell"><input type="text" size="60" onkeyup="render(this.value)" /></td>
    </tr>
  </table>
</div>

<div id="list">
</div>


<div class="footer">generated by <a href="http://search.cpan.org/perldoc?Pod/ProjectDocs">Pod::ProjectDocs</a></div>
</body>
</html>
