<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en">
<head>
	<title></title>
<link rel="stylesheet" type="text/css" href="podstyle.css" /><meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
  <script type="text/javascript">

  function toggleCode( id ) {
    if ( document.getElementById )
      elem = document.getElementById( id );
    else if ( document.all )
      elem = eval( "document.all." + id );
    else
      return false;

    elemStyle = elem.style;

    if ( elemStyle.display != "block" ) {
      elemStyle.display = "block"
    } else {
      elemStyle.display = "none"
    }

    return true;
  }
  document.writeln( "<style type=\"text/css\">div.method-source-code { display: none }</style>" )
  </script></head>
<body>
<div class="box">
  <h1 class="t1">GAL</h1>
  <table>
    <tr>
      <td class="label">Description</td>
      <td class="cell">GAL - The Genome Annotation Library</td>
    </tr>
  </table>
</div>
<div class="path">
  <a href="index.html">GAL</a> &gt; Perl Modules &gt;
  Manual
</div>
<div>
<a href="src/Manual.pm">Source</a>
</div>


<div class="pod">
<!-- INDEX START -->
<h3 id="TOP">Index</h3>
<ul>
	<li><a href="#Summary">Summary</a></li>
	<li><a href="#Synopsis">Synopsis</a></li>
	<li><a href="#Description">Description</a></li>
	<li><a href="#Storage">Storage</a></li>
	<li><a href="#Class_Structure">Class Structure</a></li>
	<li><a href="#Parsers">Parsers</a></li>
	<li><a href="#Using_GAL">Using GAL</a></li>
	<li><a href="#GAL_Classes">GAL Classes</a></li>
</ul>
<hr />
<!-- INDEX END -->

<h1 id="Summary">Summary <a href="#TOP" class="toplink"><img alt="^" src="up.gif" /></a></h1>

<p>The Genome Annotation Library (GAL) is a object-oriented Perl
programming library (and a growing collection of associated
scripts) that is designed to provide a simple and intuitive
interface for accomplishing Sequence Ontology empowered analyses of
genomic annotation data.</p>

<h1 id="Synopsis">Synopsis <a href="#TOP" class="toplink"><img alt="^" src="up.gif" /></a></h1>

<p>A very simple GAL script that infers introns for mRNAs is seen here:</p>

<pre>  use GAL::Annotation;
  my $features = GAL::Annotation-&gt;new($gff_file, $fasta_file)-&gt;features;

  # Search for mRNAs and get an iterator;
  my $mrnas = $features-&gt;search({type =&gt; 'mRNA'});
  print STDERR 'Processing ' . $mrnas-&gt;count . ' mRNA features';
  # Iterate mRNAs
  while (my $mrna = $mrnas-&gt;next) {
    # Infer (if not existing) and return iterator
    my $introns = $mrna-&gt;introns;
    while (my $intron = $introns-&gt;next) {
      # Use the intron object for access to data
      my $id    = $intron-&gt;feature_id;
      my ($start, $end) = $intron-&gt;get_values(qw(start end));
      print $intron-&gt;fasta if $intron-&gt;gc_content &gt; 50;
      print STDERR join ', ', ($feature_id,
			       $intron-&gt;locus,
			       $intron-&gt;gc_content
			      );
    }
  }

</pre><h1 id="Description">Description <a href="#TOP" class="toplink"><img alt="^" src="up.gif" /></a></h1>

<p>GAL provides a collection of object classes for sequence features
who's inheritance structure is based on the is_a releationships
between sequence features as described in the Sequence Ontology
(e.g. an mRNA is_a type of transcript and can thus inherit it's
methods).  In addition feature classes provide methods which will
retrieve iterators for child features as described by the SO part_of
relationships and captured by the GFF3 Parent attribute. For
example, transcript objects can return iterators for the exons which
are 'part_of' them.  mRNA objects inherit this behavior to return
exon iterators from the transcript class and in addition can return
an iterator for the CDS features that are 'part_of' them.</p>

<h1 id="Storage">Storage <a href="#TOP" class="toplink"><img alt="^" src="up.gif" /></a></h1>

<p>GAL uses an SQLite relational database backend to store parsed
sequence features.  This allows flexible searching of sequence
features and their attributes.  Results of these searches are
returned as iterators, that, in addition to iterating over the
returned features, provide useful methods for summarizing details of
their lists.  This flexibility in searching and random access to the
data comes at the cost of additional time needed to load, index and
query the database.  Sometimes, all that you need is the ability to
parse and iterate over the entire file one feature at a time.  In
these cases the parser classes can be used directly, as described
below, and features will be rapidly parsed and returned (in a simple
hash sturcture rather than feature objects).</p>

<h1 id="Class_Structure">Class Structure <a href="#TOP" class="toplink"><img alt="^" src="up.gif" /></a></h1>

<p>GAL uses DBIx::Class to provide sequence feature objects from the
relational database backend.</p>

<h1 id="Parsers">Parsers <a href="#TOP" class="toplink"><img alt="^" src="up.gif" /></a></h1>

<p>While GAL is specifically designed for parsing GFF3 as the input
format for seuqence features, an abstract Parser class allows
straight-forward subclassing such that any input file, where the
contents can be converted into genome locatable sequence features,
can can be parsed if an appropriate parser subclass is written.  In
addition to the default ability to parse GFF3, several other parser
classes are provided with GAL, and new classes are easily added.</p>

<h1 id="Using_GAL">Using GAL <a href="#TOP" class="toplink"><img alt="^" src="up.gif" /></a></h1>

<p>As described above you can use the full functionality of GAL by
allowing it to load features into a relational database or you can use
light-weight parser classes directly to rapid sequential access to the
data in hash structures.</p>

<p>In the first case a GAL::Annotation object is used to access feature
ojects:</p>

<pre>  use GAL::Annotation;
  my $features = GAL::Annotation-&gt;new($gff_file, $fasta_file)-&gt;features;
  while (my $feature = $features-&gt;next) {
      print $feature-&gt;feature_id . &quot;\n&quot;;
      print $feature-&gt;seq . &quot;\n&quot;;
  }

</pre><p>In the second case a GAL::Parser object (or subclass) is used to
access features as hash data structures:</p>

<pre>  use GAL::Parser;
  my $parser = GAL::Parser-&gt;new($gff_file);
  while (my $feature = $parser-&gt;next_feature_hash) {
      # Not $feature is a hash reference not an object!
      print $feature-&gt;{feature_id} . &quot;\n&quot;;
      # But this doesn't work $feature-&gt;{seq} - no OO sugar.
  }

</pre><h1 id="GAL_Classes">GAL Classes <a href="#TOP" class="toplink"><img alt="^" src="up.gif" /></a></h1>

<p>The following index can be used to explore the documentation in
the GAL library.</p>

<p><a href="Annotation.pm.html">GAL::Annotation</a> - The primary class used to initiate a GAL-based
annotation analysis.  Start here is you're exploring the library for
the first time and want access to the full OO feature classes and DB
backend.  Provided a GFF3 file objects of this class will load the
features into a database, index the database and can return an
iterator object based to the features in the data base.</p>

<p><a href="Base.pm.html">GAL::Base</a> - A base class providing shared utility methods for all
objects library-wide.  This class is not instantiated directly by the
user.</p>

<p><a href="Feature.pm.html">GAL::Feature</a> - Experimental and not currently in use.</p>

<p><a href="Parser.pm.html">GAL::Parser</a> - Base class for all <a href="Parser.pm.html">GAL::Parser</a> subclasses.  By
default it parsers GFF3/GVF format.  <a href="Parser.pm.html">GAL::Parser</a> is used internally
by the GAL::Annotation object to parse feature files and can also be
used directly to access features as hash references
($parser-&gt;next_feature_hash).</p>

<p><a href="Parser/gff3.pm.html">GAL::Parser::gff3</a> - A parser class used to parse GFF3 file format.
This is the default parser for the library.</p>

<p><a href="Parser/gff3_fast.pm.html">GAL::Parser::gff3_fast</a> - A fast GFF3 parser class that makes the
assumption that the incoming GFF3 file is well formated and does no
validation.  It may break in unfreindly ways if the data is not well
formated.</p>

<p><a href="Parser/ncbi_blast_tab.pm.html">GAL::Parser::ncbi_blast_tab</a> - A parser for NCBI Blast+ output that
is formated with the</p>

<p><a href="Parser/VCFv4_0.pm.html">GAL::Parser::VCFv4_0</a> - A parser for VCF v4.0 files</p>

<p><a href="Parser/basic_snp.pm.html">GAL::Parser::basic_snp</a> - A parser for SNVs in basic tab-delimited
format.</p>

<p><a href="Parser/hgmd_indel.pm.html">GAL::Parser::hgmd_indel</a> - A parser for indel flat files from HGMD.</p>

<p><a href="Parser/hgmd_snv.html.pm.html">GAL::Parser::hgmd_snv.html</a> - A parser for SNV flat files from HGMD.</p>

<p><a href="Parser/maq_cns2snp.html.pm.html">GAL::Parser::maq_cns2snp.html</a> - A parser for MAQ cns2snp output
files.</p>

<p><a href="Parser/samtools_pileup.html.pm.html">GAL::Parser::samtools_pileup.html</a> - A parser for samtools pileup
output.</p>

<p><a href="Parser/ucsc_gene_table.html.pm.html">GAL::Parser::ucsc_gene_table.html</a> - A parser for UCSC refGene (and
other) tab-delimited files.</p>

<p><a href="Parser/ucsc_kg_table.html.pm.html">GAL::Parser::ucsc_kg_table.html</a> - A parser for UCSC knownGene
tab-delimited files.</p>

<p><a href="Parser/template.html.pm.html">GAL::Parser::template.html</a> - A template to use as a starting point
for writing new GAL::Parser subclasses for sequence feature formats.</p>

<p><a href="Storage.pm.html">GAL::Storage</a> - Provides backend storage in an SQLite database.</p>

<p><a href="Storage/SQLite.pm.html">GAL::Storage::SQLite</a> - Provides backend storage in an SQLite
database.</p>

<p><a href="Schema.pm.html">GAL::Schema</a> - A class that provides basic configuration to
DBIx::Class.</p>

<p><a href="Schema/Result/Feature.pm.html">GAL::Schema::Result::Feature</a> - A DBIx::Class subclass that
describes the structure of the feature table in the database.  The
class also impliments methods that are inherited by
GAL::Schema::Result::Feature subclasses.</p>

<p><a href="Schema/Result/Feature/sequence_feature.pm.html">GAL::Schema::Result::Feature::sequence_feature</a> - A basic
sequence_feature class that provides methods for all subclasses.  The
methods are actually implimented in GAL::Schema::Result::Feature, but
inherited here so that all sequence_feature types for which no
subclass has been written will inherit all basic sequence_feature
behavior from this class.</p>

<p><a href="Schema/Result/Feature/gene.pm.html">GAL::Schema::Result::Feature::gene</a> - A class to provide gene
specific methods.</p>

<p><a href="Schema/Result/Feature/transcript.pm.html">GAL::Schema::Result::Feature::transcript</a> - A class to provide
transcript specific methods.</p>

<p><a href="Schema/Result/Feature/mrna.pm.html">GAL::Schema::Result::Feature::mrna</a> - A class to provide mRNA
specific methods.</p>

<p><a href="Schema/Result/Feature/exon.pm.html">GAL::Schema::Result::Feature::exon</a> - A class to provide exon
specific methods.</p>

<p><a href="Schema/Result/Feature/intron.pm.html">GAL::Schema::Result::Feature::intron</a> - A class to provide intron
specific methods.</p>

<p><a href="Schema/Result/Feature/cds.pm.html">GAL::Schema::Result::Feature::cds</a> - A class to provide CDS specific
methods.</p>

<p><a href="Schema/Result/Feature/protein.pm.html">GAL::Schema::Result::Feature::protein</a> - A class to provide protein
specific methods.</p>

<p><a href="Schema/Result/Feature/five_prime_utr.pm.html">GAL::Schema::Result::Feature::five_prime_utr</a> - A class to provide
five_prime_UTR specific methods.</p>

<p><a href="Schema/Result/Feature/three_prime_utr.pm.html">GAL::Schema::Result::Feature::three_prime_utr</a> - A class to provide
three_prime_UTR specific methods.</p>

<p><a href="Schema/Result/Feature/sequence_alteration.pm.html">GAL::Schema::Result::Feature::sequence_alteration</a> - A class to
provide sequence_alteration specific methods.</p>

<p><a href="Schema/Result/Feature/template.pm.html">GAL::Schema::Result::Feature::template</a> - A template class for use
in writing new GAL::Schema::Result::Feature subclasses.</p>

<p><a href="Schema/Result/Attribute.pm.html">GAL::Schema::Result::Attribute</a></p>

<p><a href="Schema/Result/Relationship.pm.html">GAL::Schema::Result::Relationship</a></p>

<p><a href="Interval.pm.html">GAL::Interval</a> - Classes that provide interval (sequence ranges)
aggregation and analysis functions for GAL.  Most of these are
currently in developement and thus not may not be functional or
stable.</p>

<p><a href="Interval/Span.pm.html">GAL::Interval::Span</a></p>

<p><a href="List.pm.html">GAL::List</a> - List aggregation and analysis methods</p>

<p><a href="List/Categorical.pm.html">GAL::List::Categorical</a> - List aggregation and analysis functions
for categorical lists</p>

<p><a href="List/Numeric.pm.html">GAL::List::Numeric</a> - List aggregation and analysis functions for
numerical lists
</p>


</div><div class="footer">generated by <a href="http://search.cpan.org/perldoc?Pod/ProjectDocs">Pod::ProjectDocs</a></div></body>
</html>
